xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,399 @@
+	// At any given time, the current balancer may be closed in the situation
+	// where there is a current balancer and a pending balancer. Thus, the
+	// swap() (close() + clearing of SubConns) operation is also guarded by this
+	// currentMu. Whether the current has been deleted or not will be checked
+	// from reading from the scToSubBalancer map, as that will be cleared in the
+	// atomic swap() operation. We are guaranteed that there can only be one
+	// balancer closed within this UpdateSubConnState() call, as multiple
+	// balancers closed would require another call to SwitchTo() which we are
+	// guaranteed won't be called concurrently. UpdateSubConnState() is
+	// different to UpdateClientConnState() and ResolverError(), as for those
+	// functions you don't update the current balancer in the situation where
+	// the current balancer could be closed (current + pending balancer
+	// populated).
+	gsb.currentMu.Lock()
+	defer gsb.currentMu.Unlock()
+	gsb.mu.Lock()
+	if gsb.closed {
+		gsb.mu.Unlock()
+		return
+	}
+	// This SubConn update will forward to the current balancer even if there is
+	// a pending present. This is because if this balancer does not forward the
+	// update, the picker from the current will not be updated with SubConns,
+	// leading to the possibility that the ClientConn constantly picks bad
+	// SubConns in the Graceful Switch period.
+	var balToUpdate balancer.Balancer
+	if gsb.balancerCurrent != nil && gsb.balancerCurrent.subconns[sc] {
+		balToUpdate = gsb.balancerCurrent
+	}
+	if gsb.balancerPending != nil && gsb.balancerPending.subconns[sc] {
+		balToUpdate = gsb.balancerPending
+	}
+	if balToUpdate == nil {
+		// SubConn belonged to a stale LB policy that has not yet fully closed.
+		gsb.mu.Unlock()
+		return
+	}
+	gsb.mu.Unlock()
+	balToUpdate.UpdateSubConnState(sc, state)
+}
+
+// Close closes any active child balancers.
+func (gsb *Balancer) Close() {
+	gsb.mu.Lock()
+	gsb.closed = true
+	currentBalancerToClose := gsb.balancerCurrent
+	gsb.balancerCurrent = nil
+	pendingBalancerToClose := gsb.balancerPending
+	gsb.balancerPending = nil
+	gsb.mu.Unlock()
+
+	if currentBalancerToClose != nil {
+		currentBalancerToClose.Close()
+	}
+	if pendingBalancerToClose != nil {
+		pendingBalancerToClose.Close()
+	}
+}
+
+// ExitIdle forwards the call to the latest balancer created.
+func (gsb *Balancer) ExitIdle() {
+	balToUpdate := gsb.latestBalancer()
+	if balToUpdate == nil {
+		return
+	}
+	// There is no need to protect this read with a mutex, as the write to
+	// .Balancer will never happen concurrently.
+	if ei, ok := balToUpdate.Balancer.(balancer.ExitIdler); ok {
+		ei.ExitIdle()
+	}
+}
+
+type balancerWrapper struct {
+	balancer.Balancer
+	gsb *Balancer
+
+	lastState balancer.State
+	subconns  map[balancer.SubConn]bool // subconns created by this balancer
+}
+
+// After this, test to make sure it works
+func (bw *balancerWrapper) Close() {
```

Comment:
Should there be a comment like "bw must not be referenced via balancerPending or balancerCurrent in gsb before Close is called"?

Reply:
Sure, added.